### PR TITLE
[WEB-Po-Admin] Task #220439: Event type list, Task#221113: [WEB-Po-Admin] Show event details on click of event, Task#220874: [WEB-Po-Admin] For add and edit open a same component

### PR DESCRIPTION
**Pull request by snehalsabade**
https://tracker.tekdi.net/issues/220439
https://tracker.tekdi.net/issues/220874
https://tracker.tekdi.net/issues/221113

### I have ensured that following `Pull Request Checklist` is taken care of before sending this PR
* [] Code is formatted as per format decided
* [] Updated acceptance criteria in tracker
* [] Updated test cases in test-cases-tracker

## Summary by CodeRabbit

- **New Features**
  - Added a form component for creating and managing "DO IDs" with validation and submission functionalities.
  - Introduced a list component to manage DO IDs with sorting, searching, pagination, and row selection.

- **Enhancements**
  - Updated the event detail view component to include edit actions and improved display of event information.

- **Bug Fixes**
  - Enhanced form handling and submission with better error handling and state management.

- **API Integrations**
  - Added new functions for fetching, creating, and updating event DO IDs to the event service.

### CODE REVIEW

File: apps/front-end/src/pages/admin/PoAdmin/DoId/View.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+              ),
+              link: "/poadmin/do-ids",
+              icon: "GroupLineIcon",
+            },
+
+            <Chip
+              textAlign="center"
+              lineHeight="15px"
+              label={id}
+            />,
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/View.js]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    } else {
+      setFormData(newData);
+      setErrors({
+        [result?.key || "name"]: {
+          __errors: [result?.message],
+        },
+      });
+    }
+    setLoading(false);
+  };
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      key: "event_type",
+      arr: result?.data?.FACILITATOR_EVENT_TYPE?.map((e) => ({
+        ...e,
+        title: t(e.title),
+      })),
+      title: "title",
+      value: "value",
+    });
+    setSchema(newSchema);
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js]: not required

File: lib/common-lib/src/services/EventService.js [hunk 4]
```diff
@@ -231,3 +231,86 @@ export const startExam = async (id, data, header = {}) => {
...
+    } else {
+      return {}
+    }
+  } catch ({ response, message }) {
+    return {
+      status: response?.status ? response?.status : 404,
+      error: response?.data?.message ? response?.data?.message : message
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply lib/common-lib/src/services/EventService.js:262]: Resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/List.js [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+              setFilter({ ...filter, limit: e, page: 1 });
+            },
+            [setFilter, filter]
+          )}
+          onChangePage={useCallback(
+            (e) => {
+              setFilter({ ...filter, page: e });
+            },
+            [setFilter, filter]
+          )}
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/List.js:281]: use to recalculating when setFilter or filter changes.

File: apps/front-end/src/pages/admin/PoAdmin/DoId/List.js [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    (row) => {
+      navigate(`/poadmin/do-ids/${row?.id}`, {
+        state: { eventData: row },
+      });
+    },
+    [navigate]
+  );
+
+  const handleProgramChange = async (selectedItem) => {
+    const data = programList.find((e) => e.id == selectedItem);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/List.js:150]: Resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                  </AdminTypo.H4>
+                ),
+              },
+            ]}
+          />
+        </VStack>
+
+        <VStack p="4" space={4}>
+          <Form
+            key={(lang, formData)}
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js]: Resolved

File: apps/front-end/src/pages/admin/PoAdmin/Address/View.js [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+              ),
+              link: "/poadmin/ips",
+              icon: "GroupLineIcon",
+            },
+
+            <Chip
+              textAlign="center"
+              lineHeight="15px"
+              label={organisation?.id}
+            />,
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/Address/View.js]: Not required

File: apps/front-end/src/pages/admin/PoAdmin/DoId/List.js [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+                    _icon={{ size: "18px" }}
+                    name="AddLineIcon"
+                  />
+                }
+                onPress={(e) => navigate("/poadmin/do-ids/create")}
+              >
+                {t("ADD")}
+              </Button>
+            )}
+          ></Menu>
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/List.js:228]: Resolved

File: apps/front-end/src/pages/admin/PoAdmin/Form.js [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+      //   format: "FileUpload",
+      // },
+      // doc_quarterly_id: {
+      //   type: "string",
+      //   label: "MONTHLY_UTILIZATION",
+      //   document_type: "camp",
+      //   document_sub_type: "consent_form",
+      //   format: "FileUpload",
+      // },
+    },
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/Form.js]: Resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  const location = useLocation();
+  const [schema, setSchema] = useState(Schema);
+  const [formData, setFormData] = useState({});
+  const [event, setEvent] = useState(location.state?.eventData || undefined);
+
+  useEffect(async () => {
+    const result = await enumRegistryService.listOfEnum();
+    let newSchema = { ...Schema };
+
+    if (event && event.id) {
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:42]: Not required
Sometimes event get undefined, that is why I have added event && event.id

File: apps/front-end/src/pages/admin/PoAdmin/Address/List.js [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+      setFilter({
+        ...filter,
+        order_by: { ...(filter?.order_by || {}), [column?.sortField]: sort },
+      });
+    }
+  };
+
+  const handleSearch = (e) => {
+    setFilter((item) => ({ ...item, search: e.nativeEvent.text, page: 1 }));
+  };
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/Address/List.js]: search by event-type
event_type: e.nativeEvent.text

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      setErrors({
+        [result?.key || "name"]: {
+          __errors: [result?.message],
+        },
+      });
+    }
+    setLoading(false);
+  };
+
+  useEffect(async () => {
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js]: Not required

File: apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  }, []);
+
+  const onChange = async (e, id) => {
+    const newData = e.formData;
+    if (newData?.state) {
+      setSelectedProgramId({
+        program_id: newData?.state,
+      });
+    }
+  };
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/EditDoId/Form.js]: resolved

File: lib/common-lib/src/services/EventService.js [hunk 15]
```diff
@@ -231,3 +231,86 @@ export const startExam = async (id, data, header = {}) => {
...
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply lib/common-lib/src/services/EventService.js]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/List.js [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+        program_id: obj?.id,
+        name: obj?.name,
+        state_name: obj?.state?.state_name,
+      };
+      setSelectedProgramId(defaultData);
+      setFilter({ ...filter, program_id: obj?.id });
+    } else {
+      setFilter({ ...filter, program_id: localData?.program_id });
+    }
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/List.js:119]: not required

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+      key: "event_type",
+      arr: result?.data?.FACILITATOR_EVENT_TYPE?.map((e) => ({
+        ...e,
+        title: t(e.title),
+      })),
+      title: "title",
+      value: "value",
+    });
+    setSchema(newSchema);
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js]: not required

File: lib/common-lib/src/services/EventService.js [hunk 18]
```diff
@@ -231,3 +231,86 @@ export const startExam = async (id, data, header = {}) => {
...
+      {
+        headers
+      }
+    )
+
+    if (result?.data) {
+      return result?.data?.data
+    } else {
+      return {}
+    }
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply lib/common-lib/src/services/EventService.js:255]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+  const location = useLocation();
+  const [schema, setSchema] = useState(Schema);
+  const [formData, setFormData] = useState({});
+  const [event, setEvent] = useState(location.state?.eventData || undefined);
+
+  useEffect(async () => {
+    const result = await enumRegistryService.listOfEnum();
+    let newSchema = { ...Schema };
+
+    if (event && event.id) {
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:42]: removed

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+      result = await eventService.createDoId(newData);
+    }
+    if (result.error) {
+      setFormData(newData);
+      setErrors({ [result?.key || "name"]: { __errors: [result?.message] } });
+      return;
+    }
+    navigate("/poadmin/do-ids");
+    setLoading(false);
+  };
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:102]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+      key: "event_type",
+      arr: result?.data?.FACILITATOR_EVENT_TYPE?.map((e) => ({
+        ...e,
+        title: t(e.title),
+      })),
+      title: "title",
+      value: "value",
+    });
+    setSchema(newSchema);
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:71]: not required

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:71]: NOT REQUIRED

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+  const location = useLocation();
+  const [schema, setSchema] = useState(Schema);
+  const [formData, setFormData] = useState({});
+  const [event, setEvent] = useState(location.state?.eventData || undefined);
+
+  useEffect(async () => {
+    const result = await enumRegistryService.listOfEnum();
+    let newSchema = { ...Schema };
+
+    if (event && event.id) {
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:42]: while remove getting error

**Review by sagarkoshti1990 — CHANGES_REQUESTED** [review]

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+      key: "event_type",
+      arr: result?.data?.FACILITATOR_EVENT_TYPE?.map((e) => ({
+        ...e,
+        title: t(e.title),
+      })),
+      title: "title",
+      value: "value",
+    });
+    setSchema(newSchema);
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:72]: not required

File: apps/front-end/src/pages/admin/PoAdmin/DoId/View.js [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+
+  useEffect(async () => {
+    const data = await eventService.getOneDoIdDetails({ id });
+    setDoId(data?.data);
+  }, []);
+
+  const handleEditButton = () => {
+    const step = "edit";
+    navigate(`/poadmin/do-ids/${id}/edit`);
+  };
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/View.js:30]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/View.js [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+  const { t } = useTranslation();
+  const { id } = useParams();
+  const [doId, setDoId] = useState();
+
+  const navigate = useNavigate();
+
+  useEffect(async () => {
+    const data = await eventService.getOneDoIdDetails({ id });
+    setDoId(data?.data);
+  }, []);
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/View.js:25]: resolved

File: apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+      setLoading(false);
+    } catch (error) {
+      console.error("Failed to submit form:", error);
+      setErrors({ submit: { __errors: ["An unexpected error occurred"] } });
+    }
+  };
+
+  if (!schema) {
+    return <div>Loading...</div>; // Or any loading indicator
+  }
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply apps/front-end/src/pages/admin/PoAdmin/DoId/Form.js:112]: resolved

File: lib/common-lib/src/services/EventService.js [hunk 28]
```diff
@@ -231,3 +231,101 @@ export const startExam = async (id, data, header = {}) => {
...
+    )
+    if (result?.data) {
+      return result?.data
+    } else {
+      return {}
+    }
+  } catch (e) {
+    return handleResponseException(e)
+  }
+}
```
[the remark this answers is not in the document]
Comment (snehalsabade) [reply lib/common-lib/src/services/EventService.js:257]: resolved

**Pull request merged**